MenuItem: doesn't work when used inside composite component with a form

When using a composite component which has a h:form inside and using <cc:renderFacet> to put content inside, if you use <p:splitButton> with <p:menuItem> it results in an exception:
```
Caused by: javax.faces.FacesException: MenuItem must be inside a form element
    at org.primefaces.renderkit.MenuItemAwareRenderer.encodeOnClick (MenuItemAwareRenderer.java:81)
    at org.primefaces.component.splitbutton.SplitButtonRenderer.encodeMenuItem (SplitButtonRenderer.java:339)
    at org.primefaces.component.splitbutton.SplitButtonRenderer.encodeElements (SplitButtonRenderer.java:292)
    at org.primefaces.component.splitbutton.SplitButtonRenderer.encodeMenu (SplitButtonRenderer.java:263)
    at org.primefaces.component.splitbutton.SplitButtonRenderer.encodeMarkup (SplitButtonRenderer.java:103)
    at org.primefaces.component.splitbutton.SplitButtonRenderer.encodeEnd (SplitButtonRenderer.java:77)
    at javax.faces.component.UIComponentBase.encodeEnd (UIComponentBase.java:675)
    at javax.faces.component.UIComponentBase.encodeAll (UIComponentBase.java:555)
    at org.apache.myfaces.renderkit.html.HtmlCompositeFacetRenderer.encodeChildren (HtmlCompositeFacetRenderer.java:80)
    at javax.faces.component.UIComponentBase.encodeChildren (UIComponentBase.java:644)
    at javax.faces.component.UIComponentBase.encodeAll (UIComponentBase.java:542)
    at javax.faces.component.UIComponentBase.encodeAll (UIComponentBase.java:551)
    at javax.faces.component.UIComponentBase.encodeAll (UIComponentBase.java:551)
    at org.apache.myfaces.renderkit.html.HtmlCompositeComponentRenderer.encodeChildren (HtmlCompositeComponentRenderer.java:66)
    at javax.faces.component.UIComponentBase.encodeChildren (UIComponentBase.java:644)
    at javax.faces.component.UIComponentBase.encodeAll (UIComponentBase.java:542)
    at javax.faces.component.UIComponentBase.encodeAll (UIComponentBase.java:551)
    at javax.faces.component.UIComponentBase.encodeAll (UIComponentBase.java:551)
    at org.apache.myfaces.view.facelets.FaceletViewDeclarationLanguage.renderView (FaceletViewDeclarationLanguage.java:1891)
    at org.apache.myfaces.application.ViewHandlerImpl.renderView (ViewHandlerImpl.java:313)
    at javax.faces.application.ViewHandlerWrapper.renderView (ViewHandlerWrapper.java:58)
    at javax.faces.application.ViewHandlerWrapper.renderView (ViewHandlerWrapper.java:58)
    at org.apache.myfaces.lifecycle.RenderResponseExecutor.execute (RenderResponseExecutor.java:116)
    at org.apache.myfaces.lifecycle.LifecycleImpl.render (LifecycleImpl.java:267)
    at javax.faces.webapp.FacesServlet.service (FacesServlet.java:200)
    at org.eclipse.jetty.servlet.ServletHolder$NotAsync.service (ServletHolder.java:1459)
    at org.eclipse.jetty.servlet.ServletHolder.handle (ServletHolder.java:799)
    at org.eclipse.jetty.servlet.ServletHandler$ChainEnd.doFilter (ServletHandler.java:1631)
    at org.eclipse.jetty.servlet.ServletHandler.doHandle (ServletHandler.java:548)
    at org.eclipse.jetty.server.handler.ScopedHandler.handle (ScopedHandler.java:143)
    at org.eclipse.jetty.security.SecurityHandler.handle (SecurityHandler.java:618)
    at org.eclipse.jetty.server.handler.HandlerWrapper.handle (HandlerWrapper.java:127)
    at org.eclipse.jetty.server.handler.ScopedHandler.nextHandle (ScopedHandler.java:235)
    at org.eclipse.jetty.server.session.SessionHandler.doHandle (SessionHandler.java:1624)
    at org.eclipse.jetty.server.handler.ScopedHandler.nextHandle (ScopedHandler.java:233)
    at org.eclipse.jetty.server.handler.ContextHandler.doHandle (ContextHandler.java:1434)
    at org.eclipse.jetty.server.handler.ScopedHandler.nextScope (ScopedHandler.java:188)
    at org.eclipse.jetty.servlet.ServletHandler.doScope (ServletHandler.java:501)
    at org.eclipse.jetty.server.session.SessionHandler.doScope (SessionHandler.java:1594)
    at org.eclipse.jetty.server.handler.ScopedHandler.nextScope (ScopedHandler.java:186)
    at org.eclipse.jetty.server.handler.ContextHandler.doScope (ContextHandler.java:1349)
    at org.eclipse.jetty.server.handler.ScopedHandler.handle (ScopedHandler.java:141)
    at org.eclipse.jetty.server.Dispatcher.forward (Dispatcher.java:229)
    at org.eclipse.jetty.server.Dispatcher.forward (Dispatcher.java:81)
    at org.eclipse.jetty.server.ResourceService.sendWelcome (ResourceService.java:439)
    at org.eclipse.jetty.server.ResourceService.doGet (ResourceService.java:257)
    at org.eclipse.jetty.servlet.DefaultServlet.doGet (DefaultServlet.java:447)
    at javax.servlet.http.HttpServlet.service (HttpServlet.java:687)
    at javax.servlet.http.HttpServlet.service (HttpServlet.java:790)
    at org.eclipse.jetty.servlet.ServletHolder$NotAsync.service (ServletHolder.java:1450)
    at org.eclipse.jetty.servlet.ServletHolder.handle (ServletHolder.java:799)
    at org.eclipse.jetty.servlet.ServletHandler$ChainEnd.doFilter (ServletHandler.java:1631)
    at org.eclipse.jetty.websocket.server.WebSocketUpgradeFilter.doFilter (WebSocketUpgradeFilter.java:228)
    at org.eclipse.jetty.servlet.FilterHolder.doFilter (FilterHolder.java:193)
    at org.eclipse.jetty.servlet.ServletHandler$Chain.doFilter (ServletHandler.java:1601)
    at org.eclipse.jetty.servlet.ServletHandler.doHandle (ServletHandler.java:548)
    at org.eclipse.jetty.server.handler.ScopedHandler.handle (ScopedHandler.java:143)
    at org.eclipse.jetty.security.SecurityHandler.handle (SecurityHandler.java:600)
    at org.eclipse.jetty.server.handler.HandlerWrapper.handle (HandlerWrapper.java:127)
    at org.eclipse.jetty.server.handler.ScopedHandler.nextHandle (ScopedHandler.java:235)
    at org.eclipse.jetty.server.session.SessionHandler.doHandle (SessionHandler.java:1624)
    at org.eclipse.jetty.server.handler.ScopedHandler.nextHandle (ScopedHandler.java:233)
    at org.eclipse.jetty.server.handler.ContextHandler.doHandle (ContextHandler.java:1434)
    at org.eclipse.jetty.server.handler.ScopedHandler.nextScope (ScopedHandler.java:188)
    at org.eclipse.jetty.servlet.ServletHandler.doScope (ServletHandler.java:501)
    at org.eclipse.jetty.server.session.SessionHandler.doScope (SessionHandler.java:1594)
    at org.eclipse.jetty.server.handler.ScopedHandler.nextScope (ScopedHandler.java:186)
    at org.eclipse.jetty.server.handler.ContextHandler.doScope (ContextHandler.java:1349)
    at org.eclipse.jetty.server.handler.ScopedHandler.handle (ScopedHandler.java:141)
    at org.eclipse.jetty.server.handler.ContextHandlerCollection.handle (ContextHandlerCollection.java:191)
    at org.eclipse.jetty.server.handler.HandlerCollection.handle (HandlerCollection.java:146)
    at org.eclipse.jetty.server.handler.HandlerWrapper.handle (HandlerWrapper.java:127)
    at org.eclipse.jetty.server.Server.handle (Server.java:516)
    at org.eclipse.jetty.server.HttpChannel.lambda$handle$1 (HttpChannel.java:400)
    at org.eclipse.jetty.server.HttpChannel.dispatch (HttpChannel.java:645)
    at org.eclipse.jetty.server.HttpChannel.handle (HttpChannel.java:392)
    at org.eclipse.jetty.server.HttpConnection.onFillable (HttpConnection.java:277)
    at org.eclipse.jetty.io.AbstractConnection$ReadCallback.succeeded (AbstractConnection.java:311)
    at org.eclipse.jetty.io.FillInterest.fillable (FillInterest.java:105)
    at org.eclipse.jetty.io.ChannelEndPoint$1.run (ChannelEndPoint.java:104)
    at org.eclipse.jetty.util.thread.QueuedThreadPool.runJob (QueuedThreadPool.java:883)
    at org.eclipse.jetty.util.thread.QueuedThreadPool$Runner.run (QueuedThreadPool.java:1034)
    at java.lang.Thread.run (Thread.java:834)


```
 Primefaces fails to detect that p:menuItem is  inside a form.


**Environment:**
 - PF Version: _11_
 - JSF + version: I tested on myfaces22 but I believe it happens in all implementations
 - Affected browsers: ALL_

**To Reproduce**
See attached primefaces-test project 

[primefaces-test.zip](https://github.com/primefaces/primefaces/files/7920844/primefaces-test.zip)


**Expected behavior**
Primefaces should be able to detect that the p:menuItem is inside a h:form already even though the h:form is inside the composite component

**Example XHTML**
**test.xhtml**
```html
<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml"
      xmlns:ui="http://java.sun.com/jsf/facelets"
      xmlns:f="http://java.sun.com/jsf/core"
      xmlns:p="http://primefaces.org/ui"
      xmlns:h="http://java.sun.com/jsf/html"
      xmlns:my="http://xmlns.jcp.org/jsf/composite/my"
      >

    <h:head>   
        <title>PrimeFaces Test</title>
    </h:head>
    <h:body>
        <br/><br/>

        <my:component >
            <f:facet name="content">
                <p:splitButton value="Test">
                    <p:menuitem value="Test2" />
                </p:splitButton>
            </f:facet>

        </my:component>
    </h:body>
</html>
```
**component.xhtml**
```html
<?xml version='1.0' encoding='UTF-8' ?>
<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">
<html xmlns="http://www.w3.org/1999/xhtml"
      xmlns:cc="http://xmlns.jcp.org/jsf/composite"
      xmlns:p="http://primefaces.org/ui"
      xmlns:composite="http://java.sun.com/jsf/composite"
      xmlns:h="http://xmlns.jcp.org/jsf/html"
      xmlns:f="http://xmlns.jcp.org/jsf/core">


    <cc:interface>


    </cc:interface>

   
    <cc:implementation>
        <h:form>
            <cc:renderFacet name="content" />
        </h:form>

    </cc:implementation>
</html>

--- a/primefaces/src/main/java/org/primefaces/component/commandbutton/CommandButtonRenderer.java
+++ b/primefaces/src/main/java/org/primefaces/component/commandbutton/CommandButtonRenderer.java
@@ -154,7 +154,9 @@ public class CommandButtonRenderer extends CoreRenderer {
         else {
             UIForm form = ComponentTraversalUtils.closestForm(context, button);
             if (form == null) {
-                LOGGER.log(Level.FINE, "CommandButton : \"" + clientId + "\" should be inside a form element");
+                LOGGER.log(Level.FINE, "CommandButton '" + clientId
+                            + "' should be inside a form or should reference a form via its form attribute."
+                            + " We will try to find a fallback form on the client side.");
             }
 
             request = buildNonAjaxRequest(context, button, form, null, false);

--- a/primefaces/src/main/java/org/primefaces/component/commandbutton/CommandButtonRenderer.java
+++ b/primefaces/src/main/java/org/primefaces/component/commandbutton/CommandButtonRenderer.java
@@ -24,6 +24,8 @@
 package org.primefaces.component.commandbutton;
 
 import java.io.IOException;
+import java.util.logging.Level;
+import java.util.logging.Logger;
 
 import javax.faces.FacesException;
 import javax.faces.component.UIComponent;
@@ -40,6 +42,8 @@ import org.primefaces.util.HTML;
 import org.primefaces.util.WidgetBuilder;
 
 public class CommandButtonRenderer extends CoreRenderer {
+
+    private static final Logger LOGGER = Logger.getLogger(CommandButtonRenderer.class.getName());
 
     @Override
     public void decode(FacesContext context, UIComponent component) {
@@ -150,7 +154,7 @@ public class CommandButtonRenderer extends CoreRenderer {
         else {
             UIForm form = ComponentTraversalUtils.closestForm(context, button);
             if (form == null) {
-                throw new FacesException("CommandButton : \"" + clientId + "\" must be inside a form element");
+                LOGGER.log(Level.FINE, "CommandButton : \"" + clientId + "\" should be inside a form element");
             }
 
             request = buildNonAjaxRequest(context, button, form, null, false);

--- a/primefaces/src/main/java/org/primefaces/renderkit/CoreRenderer.java
+++ b/primefaces/src/main/java/org/primefaces/renderkit/CoreRenderer.java
@@ -493,7 +493,15 @@ public abstract class CoreRenderer extends Renderer {
     protected String buildNonAjaxRequest(FacesContext context, UIComponent component, UIComponent form, String decodeParam,
                                          Map<String, List<String>> parameters, boolean submit) {
         StringBuilder request = SharedStringBuilder.get(context, SB_BUILD_NON_AJAX_REQUEST);
-        String formId = form.getClientId(context);
+
+        String submitId;
+        if (form == null) {
+            submitId = component.getClientId(context);
+        }
+        else {
+            submitId = form.getClientId(context);
+        }
+
         Map<String, Object> params = new HashMap<>();
 
         if (decodeParam != null) {
@@ -514,7 +522,7 @@ public abstract class CoreRenderer extends Renderer {
 
         //append params
         if (!params.isEmpty()) {
-            request.append("PrimeFaces.addSubmitParam('").append(formId).append("',{");
+            request.append("PrimeFaces.addSubmitParam('").append(submitId).append("',{");
 
             request.append(
                     params.entrySet().stream()
@@ -527,7 +535,7 @@ public abstract class CoreRenderer extends Renderer {
 
         if (submit) {
             Object target = component.getAttributes().get("target");
-            request.append(".submit('").append(formId).append("'");
+            request.append(".submit('").append(submitId).append("'");
 
             if (target != null) {
                 request.append(",'").append(target).append("'");

--- a/primefaces/src/main/java/org/primefaces/renderkit/MenuItemAwareRenderer.java
+++ b/primefaces/src/main/java/org/primefaces/renderkit/MenuItemAwareRenderer.java
@@ -84,7 +84,6 @@ public class MenuItemAwareRenderer extends OutcomeTargetRenderer {
                 LOGGER.log(Level.FINE, "Menu '" + menuClientId
                             + "' should be inside a form or should reference a form via its form attribute."
                             + " We will try to find a fallback form on the client side.");
-                
             }
 
             String command;

--- a/primefaces/src/main/java/org/primefaces/renderkit/MenuItemAwareRenderer.java
+++ b/primefaces/src/main/java/org/primefaces/renderkit/MenuItemAwareRenderer.java
@@ -77,15 +77,18 @@ public class MenuItemAwareRenderer extends OutcomeTargetRenderer {
         //POST
         else {
             writer.writeAttribute("href", "#", null);
+            String menuClientId = source.getClientId(context);
 
             UIForm form = ComponentTraversalUtils.closestForm(context, source);
             if (form == null) {
-                LOGGER.log(Level.FINE, "MenuItem should be inside a form element");
+                LOGGER.log(Level.FINE, "Menu '" + menuClientId
+                            + "' should be inside a form or should reference a form via its form attribute."
+                            + " We will try to find a fallback form on the client side.");
+                
             }
 
             String command;
             if (menuitem.isDynamic()) {
-                String menuClientId = source.getClientId(context);
                 Map<String, List<String>> params = menuitem.getParams();
                 if (params == null) {
                     params = new LinkedHashMap<>();

--- a/primefaces/src/main/java/org/primefaces/renderkit/MenuItemAwareRenderer.java
+++ b/primefaces/src/main/java/org/primefaces/renderkit/MenuItemAwareRenderer.java
@@ -25,8 +25,9 @@ package org.primefaces.renderkit;
 
 import java.io.IOException;
 import java.util.*;
+import java.util.logging.Level;
+import java.util.logging.Logger;
 
-import javax.faces.FacesException;
 import javax.faces.component.UIComponent;
 import javax.faces.component.UIForm;
 import javax.faces.component.behavior.ClientBehavior;
@@ -50,6 +51,7 @@ import org.primefaces.util.HTML;
 
 public class MenuItemAwareRenderer extends OutcomeTargetRenderer {
 
+    private static final Logger LOGGER = Logger.getLogger(MenuItemAwareRenderer.class.getName());
     private static final String SEPARATOR = "_";
 
     @Override
@@ -78,7 +80,7 @@ public class MenuItemAwareRenderer extends OutcomeTargetRenderer {
 
             UIForm form = ComponentTraversalUtils.closestForm(context, source);
             if (form == null) {
-                throw new FacesException("MenuItem must be inside a form element");
+                LOGGER.log(Level.FINE, "MenuItem should be inside a form element");
             }
 
             String command;

--- a/primefaces/src/main/resources/META-INF/resources/primefaces/core/core.js
+++ b/primefaces/src/main/resources/META-INF/resources/primefaces/core/core.js
@@ -144,7 +144,7 @@
          * @param {string} [target] The target attribute to use on the form during the submit process.
          */
         submit : function(formId, target) {
-            var form = PrimeFaces.getForm(parent);
+            var form = PrimeFaces.getForm(formId);
             var prevTarget;
 
             if (target) {

--- a/primefaces/src/main/resources/META-INF/resources/primefaces/core/core.js
+++ b/primefaces/src/main/resources/META-INF/resources/primefaces/core/core.js
@@ -110,7 +110,7 @@
          * @return {JQuery} the form or NULL if no form found
          */
         getClosestForm: function(id) {
-            var form = $(this.escapeClientId(id));
+            var form = $(PrimeFaces.escapeClientId(id));
             if (!form.is('form')) {
                 form = form.closest('form');
             }

--- a/primefaces/src/main/resources/META-INF/resources/primefaces/core/core.js
+++ b/primefaces/src/main/resources/META-INF/resources/primefaces/core/core.js
@@ -103,6 +103,22 @@
         },
 
         /**
+         * Gets the form by id or the closest form if the id is not a form itself.
+         * @param {string} id ID of the component to get the closest form.
+         * @return {JQuery} the form or NULL if no form found
+         */
+        getForm: function(id) {
+            var form = $(this.escapeClientId(id));
+            if (!form.is('form')) {
+                form = form.closest('form');
+            }
+            if (!form) {
+                PrimeFaces.error('Form element could not be found.');
+            }
+            return form;
+        },
+
+        /**
          * Adds hidden input elements to the given form. For each key-value pair, a new hidden input element is created
          * with the given value and the key used as the name.
          * @param {string} parent The ID of a FORM element.
@@ -110,7 +126,7 @@
          * @return {typeof PrimeFaces} This object for chaining.
          */
         addSubmitParam : function(parent, params) {
-            var form = $(this.escapeClientId(parent));
+            var form = PrimeFaces.getForm(parent);
 
             for(var key in params) {
                 form.append("<input type=\"hidden\" name=\"" + PrimeFaces.escapeHTML(key) + "\" value=\"" + PrimeFaces.escapeHTML(params[key]) + "\" class=\"ui-submit-param\"></input>");
@@ -128,7 +144,7 @@
          * @param {string} [target] The target attribute to use on the form during the submit process.
          */
         submit : function(formId, target) {
-            var form = $(this.escapeClientId(formId));
+            var form = PrimeFaces.getForm(parent);
             var prevTarget;
 
             if (target) {

--- a/primefaces/src/main/resources/META-INF/resources/primefaces/core/core.js
+++ b/primefaces/src/main/resources/META-INF/resources/primefaces/core/core.js
@@ -104,16 +104,18 @@
 
         /**
          * Gets the form by id or the closest form if the id is not a form itself.
-         * @param {string} id ID of the component to get the closest form.
+         * In AJAX we also have a fallback for the first form in DOM, this should not be used here.
+         *
+         * @param {string} id ID of the component to get the closest form or if its a form itself
          * @return {JQuery} the form or NULL if no form found
          */
-        getForm: function(id) {
+        getClosestForm: function(id) {
             var form = $(this.escapeClientId(id));
             if (!form.is('form')) {
                 form = form.closest('form');
             }
             if (!form) {
-                PrimeFaces.error('Form element could not be found.');
+                PrimeFaces.error('Form element could not be found for id: ' + id);
             }
             return form;
         },
@@ -126,7 +128,7 @@
          * @return {typeof PrimeFaces} This object for chaining.
          */
         addSubmitParam : function(parent, params) {
-            var form = PrimeFaces.getForm(parent);
+            var form = PrimeFaces.getClosestForm(parent);
 
             for(var key in params) {
                 form.append("<input type=\"hidden\" name=\"" + PrimeFaces.escapeHTML(key) + "\" value=\"" + PrimeFaces.escapeHTML(params[key]) + "\" class=\"ui-submit-param\"></input>");
@@ -144,7 +146,7 @@
          * @param {string} [target] The target attribute to use on the form during the submit process.
          */
         submit : function(formId, target) {
-            var form = PrimeFaces.getForm(formId);
+            var form = PrimeFaces.getClosestForm(formId);
             var prevTarget;
 
             if (target) {

